### PR TITLE
Pygame base app

### PR DESCRIPTION
**Pull request by takluyver**
[Pygame](https://github.com/pygame/pygame) is a Python framework for building simple games with SDL. This is a 'base app' - like the [Electron base app](https://github.com/flathub/io.atom.electron.BaseApp) it's not meant to be run directly, but rather other developers can use it as a starting point for packaging games made with Pygame. I've tested it locally with the Aliens example game included in Pygame.

I started playing with this in 2017, then left it for over a year before coming back to it yesterday, so my memory of some of the details may be hazy.

cc @illume

### CODE REVIEW

**Comment by AsciiWolf** [comment]
bot, build org.pygame.Pygame.BaseApp

**Comment by takluyver** [comment]
Are the bot results publicly visible somewhere?

**Comment by AsciiWolf** [comment]
@takluyver Yep, [here](https://flathub.org/builds/).

File: org.pygame.Pygame.BaseApp.json [hunk 1]
```diff
@@ -0,0 +1,116 @@
+{
+    "id": "org.pygame.Pygame.BaseApp",
+    "command": "/usr/bin/bash",
```
Comment (barthalion) [inline org.pygame.Pygame.BaseApp.json]: What do you need it for?

File: org.pygame.Pygame.BaseApp.json [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                {
+                    "type": "archive",
+                    "url": "https://sourceforge.net/projects/mikmod/files/libmikmod/3.3.11.1/libmikmod-3.3.11.1.tar.gz",
+                    "sha256": "ad9d64dfc8f83684876419ea7cd4ff4a41d8bcd8c23ef37ecb3a200a16b46d19"
+                }
+            ]
+        },
+        {
+            "name": "fluidsynth",    
+            "cmake": true,
```
Comment (barthalion) [inline org.pygame.Pygame.BaseApp.json]: `"buildsystem": "cmake-ninja"`

File: org.pygame.Pygame.BaseApp.json [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                    "url": "https://downloads.sourceforge.net/project/fluidsynth/fluidsynth-1.1.6/fluidsynth-1.1.6.tar.bz2",
+                    "sha256": "d28b47dfbf7f8e426902ae7fa2981d821fbf84f41da9e1b85be933d2d748f601"
+                },
+                {
+                    "type": "patch",
+                    "path": "fluidsynth-no-rawmidi.patch"
+                }
+            ]
+        },
+        "shared-modules/SDL/SDL-1.2.15.json",
```
Comment (barthalion) [inline org.pygame.Pygame.BaseApp.json]: Please move these to the beginning of modules section if possible.

File: org.pygame.Pygame.BaseApp.json [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        "shared-modules/SDL/SDL_image-1.2.12.json",
+        "shared-modules/SDL/SDL_ttf-2.0.11.json",
+        "shared-modules/smpeg/smpeg-0.4.5.json",
+        "shared-modules/SDL/SDL_mixer-1.2.12.json",
+        {
+            "name": "portmidi",
+            "cmake": true,
```
Comment (barthalion) [inline org.pygame.Pygame.BaseApp.json]: `"buildsystem": "cmake-ninja`

**Comment by takluyver** [comment]
Thanks, fixed the comments. I see the build failed because of the lack of an appstream file. AIUI that's expected for a base app, so I've copied the file from the electron baseapp to skip the check.

I don't personally have any games I plan to submit,  I've just got a slow-burning motivation to make it easier to package Python applications. Who wants to spend time on games when there are packaging tools to play with? :wink:

File: org.pygame.Pygame.BaseApp.json (hunk 1 above)
Comment (takluyver) [reply org.pygame.Pygame.BaseApp.json]: I don't think I did need it - it was probably leftover from earlier experimentation.

**Comment by TingPing** [comment]
Just as a data point this only takes ~2 minutes to build on my machine. If we don't expect a large number of apps using it and it isn't that massive of a build (like webkit) then I think it becomes much less valuable.

**Comment by bilelmoussaoui** [comment]
@takluyver If you have a list of games created using PyGame I might look at packaging them ;-)

**Comment by takluyver** [comment]
@bilelmoussaoui People add their own games to the pygame website: https://www.pygame.org/tags/all - these don't necessarily use the pygame library, but many of them do. Some of the entries for competitions like [pyweek](https://pyweek.org/) also use pygame.

I've written an example manifest for the 'Aliens' example game that's included as part of pygame:
https://github.com/takluyver/pygame-flatpak-test/blob/master/org.pygame.aliens.json

There's also a slightly bigger example called [solarwolf](https://github.com/pygame/solarwolf) that I've previously experimented with packaging, but not updated to use this work.

@TingPing I think having this in one place is valuable in terms of saving developer time rather than saving build time. There are potentially a large number of games that could use it, but Python programmers aren't likely to engage if they have to deal with all these C dependencies. Even asking people to copy this as boilerplate would be offputting. So a baseapp feels like a useful thing to have.

I'm also thinking about tooling to simplify creating flatpaks that can be expressed as (baseapp + python packages from PyPI + launcher). I think a lot of applications could fit that pattern, and with the right design it could make flatpak much more appealing to Python developers.

**Comment by barthalion** [comment]
If it takes as short time to build it as TingPing says, it's definitely makes no sense to host it as base app on Flathub. We mostly use them only if it saves a lot of time during build (qtwebkit) or set of dependencies is repeated enough times across multiple applications.

I wouldn't mind a shared module though.

**Comment by takluyver** [comment]
Base apps may only be interesting to *you* to save build time, but I maintain that it's valuable to encapsulate this somehow to allow developers building games with pygame to distribute them as flatpaks. Base apps seemed like the obvious way to do that.

Maybe we can do something similar with a shared module? My concern is that the current shared-modules setup doesn't look like it will scale smoothly to accept a large variety of components. Whereas base apps are just apps, and you've designed Flathub for people to add apps to. :-)

Also, the obvious & recommended way to use shared-modules is with git submodules. In my experience, submodules make it easy to get yourself into a mess and then provide few hints about how to fix it.

**Comment by takluyver** [comment]
If the consensus is that a shared module is the appropriate way to
encapsulate this, though, I'll look at turning the manifest into that.

**Comment by takluyver** [comment]
I've had a go at turning this into a shared module here:
https://github.com/flathub/shared-modules/pull/20

**Comment by takluyver** [comment]
Accepted as a shared module, closing here.

**Pull request closed without merging**